txnsync/profiler.go

```diff
@@ -0,0 +1,205 @@
+// Copyright (C) 2019-2021 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package txnsync
+
+import (
+	"time"
+
+	"github.com/algorand/go-algorand/logging"
+	"github.com/algorand/go-algorand/logging/telemetryspec"
+	"github.com/algorand/go-algorand/util/timers"
+)
+
+//msgp:ignore profElements
+type profElements int
+
+const (
+	profElementIdle = iota
+	profElementTxChange
+	profElementNewRounnd
+	profElementPeerState
+	profElementIncomingMsg
+	profElementOutgoingMsg
+	profElementNextOffset
+
+	// detached elements
+	profElementGetTxnsGroups
+	profElementAssembleMessage
+	profElementSendMessage
+	profElementMakeBloomFilter
+	profElementTxnsSelection
+
+	profElementLast
+	profFirstDetachedElement = profElementGetTxnsGroups
+)
+
+// The profiler struct provides profiling information regarding the main loop performance
+// characteristics. Using it provides statistics information about the recent duty cycle utilization,
+// that could be used when trying to throlle the accuracy and performance of the transaction sync.
+type profiler struct {
+	// clock used as the source clock for measurements.
+	clock timers.WallClock
+	// elements contains the elements we want to measure.
+	elements []*element
+	// log is used to report the outcome of the measuring.
+	log logging.Logger
+
+	// profile contains all the elements indices, in order of arrival. It allows us to maintain a moving window.
+	profile []int
+	// profileSum is the total amount of time tracked by the profile array.
+	profileSum time.Duration
+	// profileSpan is the max span of the array ( or - the window ) that we would like to maintain.
+	profileSpan time.Duration
+	// spanReached flag is used to indicate whether we've already reached the window or not.
+	spanReached bool
```

Comment:
no, it's not. The first time logProfile() is called, it's still false. It turns to true only once we have a "full window" of at least `profileSpan` worth of sampling time.
let's say that profileSpan = 200ms.
and let's say that we have 5 samples:
10ms -> it will still be false
150ms -> it will still be false
30ms -> it will still be false
20ms -> it will still be true
10ms -> it will still be true

so, only after the 4th sample it would set it to true.